OpenRA.Platforms.Default/OpenAlSoundEngine.cs

```diff
@@ -404,54 +412,256 @@ public void Dispose()
+
+						// Fill a buffer and queue it for playback.
+						var nextBuffer = freeBuffers.Pop();
+						AL10.alBufferData(nextBuffer, format, data, new IntPtr(count), new IntPtr(sampleRate));
+						AL10.alSourceQueueBuffers(source, new IntPtr(1), ref nextBuffer);
+
+						lock (cts)
+							if (!cts.IsCancellationRequested)
+							{
+								// Once we have at least one buffer filled, we can actually start.
+								// If streaming fell behind, the source will have stopped,
+								// we also need to play it in this case to resume audio.
+								int state;
+								AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
+								if (state == AL10.AL_INITIAL || state == AL10.AL_STOPPED)
+								{
+									// If we resume playback from the stopped state, it resets to the beginning!
+									// To avoid replaying the same audio, we need to dequeue the processed buffers first.
+									if (state == AL10.AL_STOPPED)
+										DequeueBuffers(source, freeBuffers);
+
+									AL10.alSourcePlay(source);
+								}
+							}
+
+						// Try and dequeue buffers as they become available to be reused.
+						// When the stream ends, wait for all the buffers to be processed
+						while (freeBuffers.Count < (streamEnd ? buffers.Length : 1))
+						{
+							await Task.Delay(TimeSpan.FromSeconds(BufferSizeInSecs), cts.Token).ConfigureAwait(false);
+							DequeueBuffers(source, freeBuffers);
+						}
+					}
+				}
+				catch (TaskCanceledException)
+				{
+					// Streaming has been cancelled, we'll need to perform some cleanup.
+				}
+				finally
+				{
+					// If we never actually started the source, we need to start it and then stop it.
+					// Otherwise it is left in the initial state and never returned to the source pool.
+					int state;
+					AL10.alGetSourcei(Source, AL10.AL_SOURCE_STATE, out state);
+					if (state == AL10.AL_INITIAL)
+						AL10.alSourcePlay(Source);
+
+					// Ensure the source is stopped, which will mark all buffers as processed.
+					// Dequeue these, so they can then be freed.
+					AL10.alSourceStop(Source);
+					lock (bufferDequeueLock)
+					{
+						int buffersProcessed;
+						AL10.alGetSourcei(source, AL10.AL_BUFFERS_PROCESSED, out buffersProcessed);
+						for (var i = 0; i < buffersProcessed; i++)
+						{
+							var dequeued = 0U;
+							AL10.alSourceUnqueueBuffers(source, new IntPtr(1), ref dequeued);
+						}
+					}
+
+					AL10.alDeleteBuffers(new IntPtr(buffers.Length), buffers);
+					stream.Dispose();
+				}
+			}).ContinueWith(task =>
+			{
+				Game.RunAfterTick(() =>
+				{
+					throw new Exception("Failed to stream a sound.", task.Exception);
+				});
+			}, TaskContinuationOptions.OnlyOnFaulted);
+		}
+
+		async Task<int> ReadFillingBuffer(Stream stream, byte[] buffer)
+		{
+			var offset = 0;
+			int count;
+			while (offset < buffer.Length &&
+				(count = await stream.ReadAsync(buffer, offset, buffer.Length - offset, cts.Token).ConfigureAwait(false)) > 0)
```

Comment:
A sync read blocks the threadpool thread we are currently using until the read completes. An async read releases the thread to do other useful work (i.e. other tasks) whilst the stream is being read.

It's how async is supposed to work - free up threads that would otherwise just be blocked waiting for things and not otherwise performing useful computation.